### ISSUE TITLE
Plugins not found when using the default pylint executable

### ISSUE THREAD

**Issue by simensol** [microsoft/vscode-pylint#320]
I have installed some pylint plugins in my virtual environment. When the extension is using the default pylint executable `/usr/local/py-utils/bin/pylint`, the plugins are not found. When I execute:

```bash
/usr/local/py-utils/bin/pylint --load-plugins pylint.extensions.docparams,pylint_django,custom_pylint_plugin file.py
```

I get:

```bash
Command line or configuration file:1:0: E0013: Plugin 'pylint_django' is impossible to load, is it installed ? ('No module named 'pylint_django'') (bad-plugin-value)
Command line or configuration file:1:0: E0013: Plugin 'custom_pylint_plugin' is impossible to load, is it installed ? ('No module named 'custom_pylint_plugin'') (bad-plugin-value)
```

It doesn't complain about the `pylint.extensions.docparams` plugin, so it's only the `pylint_django` and `custom_pylint_plugin` plugins installed through pip that is not found. However, if I execute:

```bash
/root/.local/share/virtualenvs/myenv/bin/pylint --load-plugins pylint.extensions.docparams,pylint_django,custom_pylint_plugin file.py
```

everything work as expected. How can I install plugins so that they are available to the default pylint executable `/usr/local/py-utils/bin/pylint`?

**Comment by karthiknadig** [comment]
@simensol Can you tell me how you installed them? as i the command you used to install them.

The problem here is basically if python is able to find the packages. How did you create this virtual environment `/root/.local/share/virtualenvs/myenv`?

Here is how I would verify if my packages are installed in the correct way:
```
> source /root/.local/share/virtualenvs/myenv/bin/activate
> python -c "import sys;print(sys.executable)"
> python -m pip list
```
Note that I am using `-m` to install here this ensures that the right pip is used and the packages go to the right place. That should list out all the plugins you want to use. If it doesn't then python is not seeing the plugin you want. Also, the `-c` to ensure that the right python is picked, in the above case it should be `/root/.local/share/virtualenvs/myenv/bin/python`

If there are missing plugins then install them like this:
```
> /root/.local/share/virtualenvs/myenv/bin/python -m pip install pylint_django 
```

@DanielNoord may be able to provide more details as this is pylint command line.

**Comment by DanielNoord** [comment]
Yes, this does indeed feel more like a pylon support question. I guess you probably did `pip install pylint` at some point. You should doe the same (so `pip install`) for the plugins if you want to me to be available to the "global" pylint executable.

However, I would highly recommend always using a `virtualenv` and installing `pylint` and any plugins there.

**Comment by simensol** [comment]
`pylint_django` and `custom_pylint_plugin` are installed using `pipenv`. I have the following (simplified) `Pipfile`:

```toml
[[source]]
name = "pypi"
url = "https://pypi.org/simple"
verify_ssl = true

[dev-packages]
# Pylint was removed when I migrated to the Pylint extension:
# pylint = "*"
custom_pylint_plugin = { editable = true, path = "packages/custom_pylint_plugin" }
pylint-django = "*"
...

[packages]
...

[requires]
python_version = "3.10"

[pipenv]
allow_prereleases = false
```

The dependencies are installed by running `python3.10 -m pipenv install --dev` from the devcontainer's [`postStartCommand`](https://code.visualstudio.com/remote/advancedcontainers/start-processes).

Previously, I also installed `pylint` through this process, but after migrating to this Pylint extension, I don't list it in my `Pipfile` anymore.

```bash
> python -c "import sys;print(sys.executable)"
> source /root/.local/share/virtualenvs/myenv/bin/activate
```

returns:

```bash
/root/.local/share/virtualenvs/myenv/bin/python

Package                       Version   Editable project location
----------------------------- --------- ----------------------------------------
asgiref                       3.6.0
astroid                       2.15.2
Babel                         2.12.1
certifi                       2022.12.7
charset-normalizer            3.1.0
click                         8.1.3
colorama                      0.4.6
coverage                      7.2.3
cssbeautifier                 1.14.7
defusedxml                    0.7.1
dill                          0.3.6
dj-database-url               1.3.0
dj-email-url                  1.0.6
Django                        4.2
django-cache-url              3.4.4
django-cors-headers           3.14.0
django-extensions             3.2.1
django-fake-model             0.1.4
django-money                  2.1.1
django-pgtrigger              4.6.0
django-reverse-admin          2.9.6
django-stubs-ext              0.8.0
django-types                  0.17.0
djangorestframework           3.14.0
djangorestframework-types     0.8.0
djlint                        1.19.17
EditorConfig                  0.12.3
environs                      9.5.0
exceptiongroup                1.1.1
execnet                       1.9.0
get-docker-secret             1.0.2
html-tag-names                0.1.2
html-void-elements            0.1.0
httpie                        3.2.1
idna                          3.4
importlib-resources           5.12.0
inclusive-django-range-fields 0.2.3
iniconfig                     2.0.0
isort                         5.12.0
jsbeautifier                  1.14.7
lazy-object-proxy             1.9.0
lorem-text                    2.1
markdown-it-py                2.2.0
marshmallow                   3.19.0
mccabe                        0.7.0
mdurl                         0.1.2
multidict                     6.0.4
packaging                     23.0
pathspec                      0.11.1
pika                          1.3.1
pip                           23.0.1
platformdirs                  3.2.0
pluggy                        1.0.0
psycopg2                      2.9.6
py-moneyed                    1.2
Pygments                      2.15.0
pylint                        2.17.2
pylint-django                 2.5.3
custom-pylint-plugin          0.0.1     /workspaces/backend/packages/custom_pylint_plugin
pylint-plugin-utils           0.7
PySocks                       1.7.1
pytest                        7.3.0
pytest-cov                    4.0.0
pytest-django                 4.5.2
pytest-xdist                  3.2.1
python-dotenv                 1.0.0
pytz                          2023.3
PyYAML                        6.0
regex                         2023.3.23
requests                      2.28.2
requests-toolbelt             0.10.1
rich                          13.3.3
setuptools                    67.6.1
six                           1.16.0
sqlparse                      0.4.3
StrEnum                       0.4.10
tomli                         2.0.1
tomlkit                       0.11.7
tqdm                          4.65.0
typing_extensions             4.5.0
Unidecode                     1.3.6
uritemplate                   4.1.1
urllib3                       1.26.15
wheel                         0.38.4
wrapt                         1.15.0
yachalk                       0.1.5
```

`pylint` is listed, but its probably as a peer dependence. Both `pylint-django` and `custom-pylint-plugin` are also listed as expected.

My current workaround is to set `"pylint.path": ["/root/.local/share/virtualenvs/ecro/bin/pylint"]` in my `.vscode/settings.json`. However, as stated in the [documentation](https://marketplace.visualstudio.com/items?itemName=ms-python.pylint) of this extension, setting the `pylint.path` option

> [...] will slow down linting, since we will have to run pylint executable every time or file save or open.

The pylint linting is very slow after migrating the this Pylint extension, so I guess this is the reason.

**Comment by karthiknadig** [comment]
And does it not work if you set the interpreter to `pylint.interpreter` to `/root/.local/share/virtualenvs/myenv/bin/python`? This should also work if you just select `/root/.local/share/virtualenvs/myenv/bin/python` as your python from the python extension.

**Comment by simensol** [comment]
Setting `pylint.interpreter` to `/root/.local/share/virtualenvs/myenv/bin/python` (and removing `pylint.path`) doesn't work—the `pylint_django` and `custom_pylint_plugin` pylint plugins are not found.

Setting `pylint.path` back to `/root/.local/share/virtualenvs/myenv/bin/pylint`fixes the issue.

PS: `python.defaultInterpreterPath` is already `/root/.local/share/virtualenvs/myenv/bin/python`.

**Comment by karthiknadig** [comment]
@simensol Can you share the full logs form Output >Pylint? It should have the `sys.path` that the LS is using. That might provide more clues on what is going on here.

**Comment by simensol** [comment]
Opening an empty `__init__.py` file:

### With `pylint.path` equals `/root/.local/share/virtualenvs/myenv/bin/pylint` (`pylint.interpreter` not set):

```bash
2023-04-17 20:52:46.691 [info] Name: Pylint
2023-04-17 20:52:46.691 [info] Module: pylint
2023-04-17 20:52:46.691 [info] Python extension loading
2023-04-17 20:52:46.691 [info] Waiting for interpreter from python extension.
2023-04-17 20:52:46.691 [info] Python extension loaded
2023-04-17 20:52:46.691 [info] Using legacy Pylint args from `python.linting.pylintArgs`
2023-04-17 20:52:46.691 [info] Using legacy Pylint args from `python.linting.pylintArgs`
2023-04-17 20:52:46.691 [info] Server run command: /root/.local/share/virtualenvs/myenv/bin/python /root/.vscode-server/extensions/ms-python.pylint-2023.4.0/bundled/tool/lsp_server.py
2023-04-17 20:52:46.691 [info] Server: Start requested.
2023-04-17 20:52:47.720 [info] CWD Server: /workspaces/backend
2023-04-17 20:52:47.721 [info] Settings used to run Server:
[
    {
        "cwd": "/workspaces/backend",
        "workspace": "file:///workspaces/backend",
        "args": [
            "--load-plugins",
            "pylint.extensions.docparams"
        ],
        "severity": {
            "convention": "Information",
            "error": "Error",
            "fatal": "Error",
            "refactor": "Hint",
            "warning": "Warning",
            "info": "Information"
        },
        "path": [
            "/root/.local/share/virtualenvs/myenv/bin/pylint"
        ],
        "interpreter": [
            "/root/.local/share/virtualenvs/myenv/bin/python"
        ],
        "importStrategy": "useBundled",
        "showNotifications": "off",
        "extraPaths": []
    }
]

2023-04-17 20:52:47.721 [info] Global settings:
{
    "cwd": "/vscode/vscode-server/bin/linux-x64/704ed70d4fd1c6bd6342c436f1ede30d1cff4710",
    "workspace": "/vscode/vscode-server/bin/linux-x64/704ed70d4fd1c6bd6342c436f1ede30d1cff4710",
    "args": [],
    "severity": {
        "convention": "Information",
        "error": "Error",
        "fatal": "Error",
        "refactor": "Hint",
        "warning": "Warning",
        "info": "Information"
    },
    "path": [],
    "interpreter": [],
    "importStrategy": "useBundled",
    "showNotifications": "off",
    "extraPaths": []
}

2023-04-17 20:52:47.725 [info] sys.path used to run Server:
   /workspaces/backend
   /root/.vscode-server/extensions/ms-python.pylint-2023.4.0/bundled/libs
   /root/.vscode-server/extensions/ms-python.pylint-2023.4.0/bundled/tool
   /usr/local/lib/python310.zip
   /usr/local/lib/python3.10
   /usr/local/lib/python3.10/lib-dynload
   /root/.local/share/virtualenvs/myenv/lib/python3.10/site-packages
   /workspaces/backend/packages/myenvtools/src
   /workspaces/backend/packages/pylint_myenv/src
   /workspaces/backend/packages/sms/src
2023-04-17 20:52:47.909 [info] /root/.local/share/virtualenvs/myenv/bin/pylint --version
2023-04-17 20:52:47.918 [info] CWD Server: /workspaces/backend
2023-04-17 20:52:49.754 [info] 
pylint 2.17.2
astroid 2.15.2
Python 3.10.10 (main, Mar  1 2023, 15:23:44) [GCC 10.2.1 20210110]


2023-04-17 20:52:49.754 [info] Version info for linter running for /workspaces/backend:
pylint 2.17.2
astroid 2.15.2
Python 3.10.10 (main, Mar  1 2023, 15:23:44) [GCC 10.2.1 20210110]

2023-04-17 20:52:49.754 [info] SUPPORTED pylint>=2.12.2
FOUND pylint==2.17.2

2023-04-17 20:52:49.795 [info] [Trace - 8:52:49 PM] Received notification 'window/logMessage'.
2023-04-17 20:52:49.796 [info] /root/.local/share/virtualenvs/myenv/bin/pylint --reports=n --output-format=json --load-plugins pylint.extensions.docparams --clear-cache-post-run=y --from-stdin /workspaces/backend/claims/__init__.py
2023-04-17 20:52:49.796 [info] [Trace - 8:52:49 PM] Received notification 'window/logMessage'.
2023-04-17 20:52:49.797 [info] CWD Server: /workspaces/backend
```

### With `pylint.interpreter` equals `/root/.local/share/virtualenvs/myenv/bin/python` (`pylint.path` not set):

```bash
2023-04-17 20:56:12.179 [info] Name: Pylint
2023-04-17 20:56:12.179 [info] Module: pylint
2023-04-17 20:56:12.179 [info] Using legacy Pylint args from `python.linting.pylintArgs`
2023-04-17 20:56:12.179 [info] Using legacy Pylint path from `python.linting.pylintPath`
2023-04-17 20:56:12.179 [info] Using legacy Pylint args from `python.linting.pylintArgs`
2023-04-17 20:56:12.179 [info] Using legacy Pylint path from `python.linting.pylintPath`
2023-04-17 20:56:12.179 [info] Server run command: /root/.local/share/virtualenvs/myenv/bin/python /root/.vscode-server/extensions/ms-python.pylint-2023.4.0/bundled/tool/lsp_server.py
2023-04-17 20:56:12.179 [info] Server: Start requested.
2023-04-17 20:56:12.471 [info] CWD Server: /workspaces/backend
2023-04-17 20:56:12.529 [info] Settings used to run Server:
[
    {
        "cwd": "/workspaces/backend",
        "workspace": "file:///workspaces/backend",
        "args": [
            "--load-plugins",
            "pylint.extensions.docparams"
        ],
        "severity": {
            "convention": "Information",
            "error": "Error",
            "fatal": "Error",
            "refactor": "Hint",
            "warning": "Warning",
            "info": "Information"
        },
        "path": [
            "/usr/local/py-utils/bin/pylint"
        ],
        "interpreter": [
            "/root/.local/share/virtualenvs/myenv/bin/python"
        ],
        "importStrategy": "useBundled",
        "showNotifications": "off",
        "extraPaths": []
    }
]

2023-04-17 20:56:12.577 [info] Global settings:
{
    "cwd": "/vscode/vscode-server/bin/linux-x64/704ed70d4fd1c6bd6342c436f1ede30d1cff4710",
    "workspace": "/vscode/vscode-server/bin/linux-x64/704ed70d4fd1c6bd6342c436f1ede30d1cff4710",
    "args": [],
    "severity": {
        "convention": "Information",
        "error": "Error",
        "fatal": "Error",
        "refactor": "Hint",
        "warning": "Warning",
        "info": "Information"
    },
    "path": [],
    "interpreter": [],
    "importStrategy": "useBundled",
    "showNotifications": "off",
    "extraPaths": []
}

2023-04-17 20:56:12.577 [info] sys.path used to run Server:
   /workspaces/backend
   /root/.vscode-server/extensions/ms-python.pylint-2023.4.0/bundled/libs
   /root/.vscode-server/extensions/ms-python.pylint-2023.4.0/bundled/tool
   /usr/local/lib/python310.zip
   /usr/local/lib/python3.10
   /usr/local/lib/python3.10/lib-dynload
   /root/.local/share/virtualenvs/myenv/lib/python3.10/site-packages
   /workspaces/backend/packages/myenvtools/src
   /workspaces/backend/packages/pylint_myenv/src
   /workspaces/backend/packages/sms/src
2023-04-17 20:56:12.578 [info] /usr/local/py-utils/bin/pylint --version
2023-04-17 20:56:12.578 [info] CWD Server: /workspaces/backend
2023-04-17 20:56:14.136 [info] 
pylint 2.16.2
astroid 2.14.2
Python 3.10.10 (main, Mar  1 2023, 15:23:44) [GCC 10.2.1 20210110]


2023-04-17 20:56:14.137 [info] Version info for linter running for /workspaces/backend:
pylint 2.16.2
astroid 2.14.2
Python 3.10.10 (main, Mar  1 2023, 15:23:44) [GCC 10.2.1 20210110]

2023-04-17 20:56:14.137 [info] SUPPORTED pylint>=2.12.2
FOUND pylint==2.16.2

2023-04-17 20:56:14.170 [info] [Trace - 8:56:14 PM] Received notification 'window/logMessage'.
2023-04-17 20:56:14.170 [info] /usr/local/py-utils/bin/pylint --reports=n --output-format=json --load-plugins pylint.extensions.docparams --clear-cache-post-run=y --from-stdin /workspaces/backend/claims/__init__.py
2023-04-17 20:56:14.170 [info] [Trace - 8:56:14 PM] Received notification 'window/logMessage'.
2023-04-17 20:56:14.170 [info] CWD Server: /workspaces/backend
2023-04-17 20:56:15.648 [info] [Trace - 8:56:15 PM] Received notification 'window/logMessage'.
2023-04-17 20:56:15.648 [info] file:///workspaces/backend/claims/__init__.py :
[
    {
        "type": "error",
        "module": "/workspaces/backend/pyproject.toml",
        "obj": "",
        "line": 1,
        "column": 0,
        "endLine": null,
        "endColumn": null,
        "path": "pyproject.toml",
        "symbol": "unrecognized-option",
        "message": "Unrecognized option found: django-settings-module",
        "message-id": "E0015"
    },
    {
        "type": "error",
        "module": "Command line or configuration file",
        "obj": "",
        "line": 1,
        "column": 0,
        "endLine": null,
        "endColumn": null,
        "path": "Command line or configuration file",
        "symbol": "bad-plugin-value",
        "message": "Plugin 'pylint_django' is impossible to load, is it installed ? ('No module named 'pylint_django'')",
        "message-id": "E0013"
    },
    {
        "type": "error",
        "module": "Command line or configuration file",
        "obj": "",
        "line": 1,
        "column": 0,
        "endLine": null,
        "endColumn": null,
        "path": "Command line or configuration file",
        "symbol": "bad-plugin-value",
        "message": "Plugin 'pylint_django.checkers.migrations' is impossible to load, is it installed ? ('No module named 'pylint_django'')",
        "message-id": "E0013"
    },
    {
        "type": "error",
        "module": "Command line or configuration file",
        "obj": "",
        "line": 1,
        "column": 0,
        "endLine": null,
        "endColumn": null,
        "path": "Command line or configuration file",
        "symbol": "bad-plugin-value",
        "message": "Plugin 'pylint_myenv' is impossible to load, is it installed ? ('No module named 'pylint_myenv'')",
        "message-id": "E0013"
    }
]

2023-04-17 20:56:15.651 [info] [Trace - 8:56:15 PM] Received notification 'textDocument/publishDiagnostics'.
```

**Comment by karthiknadig** [comment]
@simensol in the pylint path not set case the logs show that is set to:
```
"path": [
            "/usr/local/py-utils/bin/pylint"
        ],
```

for the pylint path not set case it should look like this, in the logs:
```
{
        "cwd": "/workspaces/backend",
        "workspace": "file:///workspaces/backend",
        "args": [
            "--load-plugins",
            "pylint.extensions.docparams"
        ],
        "severity": {
            "convention": "Information",
            "error": "Error",
            "fatal": "Error",
            "refactor": "Hint",
            "warning": "Warning",
            "info": "Information"
        },
        "path": [],
        "interpreter": [
            "/root/.local/share/virtualenvs/myenv/bin/python"
        ],
        "importStrategy": "useBundled",
        "showNotifications": "off",
        "extraPaths": []
    }
```

**Comment by simensol** [comment]
So any idea what's going on in my case? `pylint.path` is not set in `settings.json`, but I still get:

```json
"path": [
            "/usr/local/py-utils/bin/pylint"
        ],
```

I also tried to set `"pylint.path": []`, but the logs still returns:

```json
"path": [
            "/usr/local/py-utils/bin/pylint"
        ],
```

**Comment by karthiknadig** [comment]
@simensol Can you look at your user settings (global settings)? 
![image](https://user-images.githubusercontent.com/3840081/232820724-974e9ecd-1723-4426-abeb-a55485ca8253.png)

Basically you have set `pylint` path for the python extension, this extension will use that.

**Comment by simensol** [comment]
Removing `pylint.path` from `settings.json` gives the following in the logs:

```json
"path": [
            "/usr/local/py-utils/bin/pylint"
        ],
```

However, `pylint.path` is not set in either _User_, _Remote_ or _Workspace_:

<img width="475" alt="Notification_Center" src="https://user-images.githubusercontent.com/4491653/233064276-93597e51-f6d2-481e-90e5-021530ebee36.png">
<img width="480" alt="Settings_—_backend__Dev_Container__backend___desktop-linux_" src="https://user-images.githubusercontent.com/4491653/233064449-457cd219-018f-49d1-aebe-937619466bd3.png">
<img width="475" alt="Settings_—_backend__Dev_Container__backend___desktop-linux_" src="https://user-images.githubusercontent.com/4491653/233064535-99290797-2541-4eea-bdcb-968332b34fa8.png">

**Comment by simensol** [comment]
I had removed all `pylint` settings from my `.devcontainer.json`, including `python.linting.pylintPath`. However, if I open `/root/.vscode-server/data/Machine/settings.json`, the `python.linting.pylintPath` is still present and set to `/usr/local/py-utils/bin/pylint`.

Even if I empty `/root/.vscode-server/data/Machine/settings.json` and rebuild my container, the `"python.linting.pylintPath": "/usr/local/py-utils/bin/pylint"` setting reappears.

A workaround is to set `"python.linting.pylintPath": ""` in `.devcontainer.json`. Then, the logs report:

```json
"path": [],
```

In other words, this seems to be a devcontainer related issue and not an issue with this pylint extension itself.

**Comment by karthiknadig** [comment]
@simensol Thanks for digging into this. Useful to know. Can I close this issue?

/cc @brettcannon is this something we can file an issue on?

**Comment by karthiknadig** [comment]
This is a dev container thing. Nothing to open on pylint :)

**Comment by DanielNoord** [comment]
> This is a dev container thing. Nothing to open on pylint :)

👍 

And thanks for the initial tag. We appreciate the work and triaging you're doing in this repository. Feel free to keep tagging for upstream involvement 😄

**Comment by brettcannon** [comment]
> is this something we can file an issue on?

It depends on what dev container and features are used (e.g., the kitchen sink image specifies the path to Pylint since it's installed by default, but is explicitly turned off). So my suspicion is the setting having a value by default is on purpose and expected, but linting for Pylint is turned off for newer instances of the image:

https://github.com/devcontainers/images/blob/18cefbc65ecd9e504b8835d077d17510727b75e2/src/universal/.devcontainer/devcontainer.json#L121-L122
